pretraining/utils/dataset_utils.py

```diff
@@ -1147,7 +1157,7 @@ def __init__(
                 )
 
         # Fetch logical shard sampling stats
-        self.n_docs_remaining = [len(d.docset) for d in self.data]
+        self.n_docs_remaining = [len(d.docset) if isinstance(d, _With_Docset) else 0 for d in self.data]
```

Comment:
I'm not sure if defaulting to `0` here works ok when `docset` is not there. If `docset` MUST be there perhaps we need to do something different with the classes to ensure only the correct types of `*Dataset` can be used here.